cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -18,16 +18,21 @@ private newtype TIRDataFlowNode =
   TOperandNode(Operand op) or
   TVariableNode(Variable var) or
   TAddressNodeStore(Instruction i) {
-    // Any instruction that is used as an address to a store
-    addressFlowInstrRTC(i, getDestinationAddress(_))
-    or
-    // Or any instruction that is used for an address to a load that needs an address node
-    exists(LoadInstruction load |
-      exists(TAddressNodeStore(load)) and
-      addressFlowInstrRTC(i, load.getSourceAddress())
+    exists(Instruction store | usedForStore(store) |
+      addressFlowInstrWithLoads(i, store) and
+      addressFlowInstrWithLoads(store, getDestinationAddress(_))
+      or
+      addressFlowInstrWithLoads(store, i) and
+      addressFlowInstrWithLoads(i, getDestinationAddress(_))
     )
   } or
-  TAddressNodeRead(Instruction i) { addressFlowInstrRTC(i, getSourceAddress(_)) }
+  TAddressNodeRead(Instruction i) {
+    exists(Instruction read | usedForRead(read) |
+      addressFlowInstrRTC(i, read)
+      or
+      addressFlowInstrRTC(read, i)
```

Comment:
This case isn't in the old definition. What's it used for? Read steps for an address computation that gets used as a function argument?

Reply:
It's really a refinement of the old restriction so that we generate fewer `TAddressNodeRead`. Previously, _any_ address that flowed to a `Load` or `ReadSideEffectInstruction` was given a `TAddressNodeRead` regardless of whether or not it ended up being used for read steps. But now we only include those addresses where it's possible to perform a read step. At first I also thought it would be enough to have:
```codeql
exists(Instruction read | usedForRead(read) |
  addressFlowInstrRTC(i, read)
}
```
which says that only those addresses that flow to a read step is given a `TAddressNodeRead`. But we also want to have a `TAddressNodeRead` for the conversion in this example:
```cpp
struct S {
  int x, y;
};

void test() {
  S s;
  sink(*(long*)&s.x);
}
```
where there are a couple of address steps that don't generate a read step:
```
r17_2(glval<S>)       = VariableAddress[s]    :
r17_3(glval<int>)     = FieldAddress[x]       : r17_2
r17_4(int *)          = CopyValue             : r17_3
r17_5(long *)         = Convert               : r17_4
r17_6(long)           = Load[?]               : &:r17_5, ~m16_2
```

here, we will have a read step from the `TAddressNodeRead(r17_2)` to `TAddressNodeRead(r17_3)`, but we're still not at the `LoadInstruction` (which is how we actually leave a sequence of `TAddressNodeRead`s and go back to flowing through `Instruction`s and `Operand`s).

So to actually ensure that `r17_4` and `r17_5` are assigned a `TAddressNodeRead` we also need to include the instructions that receive flow from a read step, which ends up looking like this:
```codeql
exists(Instruction read | usedForRead(read) |
  addressFlowInstrRTC(i, read)
  or
  addressFlowInstrRTC(read, i)
}
```